### PR TITLE
Haohui - Gives the user permission to add new user permissions role

### PR DESCRIPTION
**Pull request by Lin1404**
# Description
image

## Related PRS (if any):
This frontend PR is related to the #546 backend PR.
…

## How to test:
1. check into current branch
2. do npm run build and npm run start to run this PR locally
3. log as owner user call user A
4. go to dashboard→ Other Links→ Permissions Management→ Manage User Permissions
5. input another user name, call user B
6. find the “Add New User Permissions Role”, click the add button
7. scroll down and click submit button
8. log out and login as user B(if already logged in, make sure to log out and log back in)
9. go to dashboard→ Other Links→ Permissions Management→ Add New Role
10. create new role, input the role name, select permissions, then submit
11. the new role show in the "user roles" table

## Screenshots or videos of changes:

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/78375725/f9e5d3bb-9392-422c-9ae7-9ba931ed6b8c

## Note:
Include the information the reviewers need to know.

### CODE REVIEW

**Review by wantingxu7 — DISMISSED** [review]
Hi, I followed the directions and assigned permission to a volunteer role and that volunteer can create a new role. Great work!

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/105323719/fc423a78-051c-4bfd-b63a-a9e8eb3407e8

**Review by JYXiao-2021 — APPROVED** [review]
Hi, I test this pr and it works as expected. Well done!

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/49748716/32f543be-1eb4-45fd-9b34-81d260f3bc78

**Review by navneeeth — DISMISSED** [review]
Hello, @Lin1404! I tested your changes and they work as intended. Logged in as an Owner user and navigated to `Dashboard-> Other Links-> Permissions Management-> Manage User Permissions` and added the "Add New user Permissions Role" to a Volunteer user. Then logged in as the above Volunteer user and navigated to `Dashboard-> Other Links-> Permissions Management-> Add New Role` and created a new test role,  added some permissions, and then hit Submit. The newly created role shows up as intended in the `User Roles` tab. Good implementation!


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/44264834/433cd846-7d4e-4abd-a5ab-708b77660ea9


By the way, it will be helpful for reviewers if you can link the backend PR to the text that mentions it in the 'Related PRs' section in the Description. Thanks.

**Review by vikrambadhan — APPROVED** [review]
Hi, I tested your PR and it works perfectly! Great Job!

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/85417946/45b79c36-f055-489a-89cf-b40daf34c3de

**Review by Shiwani99 — APPROVED** [review]
Hi,
Great work. I test your PR. It works as expected.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/78906820/5425547a-e90c-4532-8003-8a7d054224ac

**Review by Bhagyashree444 — CHANGES_REQUESTED** [review]
From owners account I was able to add "Add New User Permissions Role" for Manager but when logged in as manager and was while trying to add "Add New Role", the application crashed.
Screenshot (140)
Screenshot (142)
Screenshot (143)

**Comment by Lin1404** [comment]
> From owners account I was able to add "Add New User Permissions Role" for Manager but when logged in as manager and was while trying to add "Add New Role", the application crashed. Screenshot (140) Screenshot (142) Screenshot (143)

Hi @Bhagyashree444 , 

Thank you for the edge case report. I have located this bug is not from my change, but I have fixed it now. Please check the row 46,47 in 'src/components/PermissionsManagement/PermissionsManagement.jsx' file.

Thanks for the edge case report again.

**Review by Bhagyashree444 — APPROVED** [review]
Thanks for making the changes. It Works as expected.

**Review by StrawberryCalpico — CHANGES_REQUESTED** [review]
Hi! I just reviewed your PR, it seems like there's a conflict with the current branch, so that I can't find the add new user permission role.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/114443664/d0f59139-a7c6-4114-ab9d-70ad1afdca4b

File: src/components/Header/Header.jsx [hunk 1]
```diff
@@ -218,7 +221,7 @@ export const Header = props => {
                         </DropdownItem>
                       </>
                     ) : null}
-                    {(canPutRole || canManageUser) && (
+                    {((canPutRole || canManageUser) || canPostRole) && (
```
Comment (nathanah) [inline src/components/Header/Header.jsx:224]:
```suggestion
                    {(canPutRole || canManageUser || canPostRole) && (
```
Extra () unneeded.

**Review by malikjahanzaib — CHANGES_REQUESTED** [review]
Hi @Lin1404, it seems this branch has some conflicts. I have checked out both FE and BE branches. When I try to go to "dashboard→ Other Links→ Permissions Management→ Manage User Permissions" from an owner account, I search for the user to assign the permission but there appears no “Add New User Permissions Role”. Below is the test recording.

Test rec:


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/46221334/fb9cd6a6-499b-4d71-90fb-d151b0e47c3e

**Review by Cloid — CHANGES_REQUESTED** [review]
Hi there.

I tried enabling the "Add New User Permissions Role" but I couldn't find it within the User Profile permission management setting for a volunteer user:
Screenshot 2023-10-21 153352

**Review by ChirayuRai — CHANGES_REQUESTED** [review]
Hey! I ran this PR on my local machine, and it seems there is no option to add new user permissions role. I have a video attached to show what I mean


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/15188751/bb025d9b-d7d7-456e-9ad0-24d8479f067a

File: src/components/PermissionsManagement/PermissionsManagement.jsx [hunk 2]
```diff
@@ -40,10 +40,11 @@ function PermissionsManagement({ getAllRoles, roles, auth, getUserRole, userProf
     <div key={`${role}+permission`} className="permissions-management">
       <h1 className="permissions-management__title">User Roles</h1>
       <div key={`${role}_header`} className="permissions-management__header">
-        {userProfile?.role === 'Owner' &&
-          <div key={`${role}_name`} className="role-name-container">
+      {(userProfile?.role === 'Owner' ||
+      userProfile?.permissions?.frontPermissions.some(permission => permission === 'postRole')) &&
```
Comment (nathanah) [inline src/components/PermissionsManagement/PermissionsManagement.jsx:44]: Shouldn't have logic based on hard-coded roles. More than just owners can create roles.

```suggestion
      {props.hasPermission('postRole') &&
```

File: src/components/PermissionsManagement/PermissionsManagement.jsx [hunk 3]
```diff
@@ -68,10 +69,14 @@ function PermissionsManagement({ getAllRoles, roles, auth, getUserRole, userProf
...
             userProfile?.permissions?.frontPermissions.some(permission => permission === 'putUserProfilePermissions') ||
+            userProfile?.permissions?.frontPermissions.some(permission => permission === 'postRole') ||
             userProfile?.role === 'Owner'
             ) && (
           <div className="buttons-container">
-            {userProfile?.role === 'Owner' &&
+            {(
+            userProfile?.permissions?.frontPermissions.some(permission => permission === 'postRole') ||
+            userProfile?.role === 'Owner'
+            ) &&
```
Comment (nathanah) [inline src/components/PermissionsManagement/PermissionsManagement.jsx:79]:
```suggestion
            {props.hasPermission('postRole') &&
```

File: src/components/PermissionsManagement/PermissionsManagement.jsx
Comment (nathanah) [inline src/components/PermissionsManagement/PermissionsManagement.jsx:1]: Import hasPermission and add it to mapDispatchToProps.

**Review by nathanah — CHANGES_REQUESTED** [review]
Should use hasPermission instead of custom logic.

**Comment by palakgosalia** [comment]
Hey, I reviewed your code and tried to test it out but I couldn't see the permission in the permissions tab. Please let me know if I need to do anything else. Thankyou. 
Attaching a video of the manage user permissions page

 

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/104417272/2ccb4168-7d71-4bdc-b16e-f15d556b3f1a

**Review by Alforoan — CHANGES_REQUESTED** [review]
As you can see by the video, I was not able to find the specific role.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/118868270/1ed2f441-3dcc-413d-ae41-58fca24a008f

**Review by KH07 — COMMENTED** [review]
I checked out the current branch, ran npm install, and started the app locally. Logging in as owner User A, in Other Links -> Permission Management -> Manager User Permissions, selected User B but was not able to find the ‘Add New User Permission Role’ option.
PR#1307

**Pull request closed without merging**